In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/house-prices-advanced-regression-techniques/sample_submission.csv
/kaggle/input/house-prices-advanced-regression-techniques/data_description.txt
/kaggle/input/house-prices-advanced-regression-techniques/train.csv
/kaggle/input/house-prices-advanced-regression-techniques/test.csv


In [2]:
# imports
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.metrics import mean_absolute_error
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from xgboost import XGBRegressor
from sklearn.inspection import permutation_importance

In [3]:
# options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [4]:
# let's read in the data and take a look
X_full = pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/train.csv', index_col='Id')
X_test_full = pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/test.csv', index_col='Id')
X_full.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
Id                                                                     
1          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
2          Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
3          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
4          Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
5          Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

   Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
Id                                                                       
1        Norm     1Fam     2Story            7            5       2003   
2        Norm     1Fam     1Story            6            8       1976   
3        Norm     1Fam     2Story            7            5       2001   
4        Norm     1Fam     2Story            7            5       1915   
5        Norm     1Fam     2Story            8            5       2000   

    YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                       
1           2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
2           1976     Gable  CompShg     MetalSd     MetalSd        NaN   
3           2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4           1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
5           2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

    MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
Id                                                                             
1        196.0        Gd        TA      PConc       Gd       TA           No   
2          0.0        TA        TA     CBlock       Gd       TA           Gd   
3        162.0        Gd        TA      PConc       Gd       TA           Mn   
4          0.0        TA        TA     BrkTil       TA       Gd           No   
5        350.0        Gd        TA      PConc       Gd       TA           Av   

   BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
Id                                                                             
1           GLQ         706          Unf           0        150          856   
2           ALQ         978          Unf           0        284         1262   
3           GLQ         486          Unf           0        434          920   
4           ALQ         216          Unf           0        540          756   
5           GLQ         655          Unf           0        490         1145   

   Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
Id                                                                             
1     GasA        Ex          Y      SBrkr       856       854             0   
2     GasA        Ex          Y      SBrkr      1262         0             0   
3     GasA        Ex          Y      SBrkr       920       866             0   
4     GasA        Gd          Y      SBrkr       961       756             0   
5     GasA        Ex          Y      SBrkr      1145      1053             0   

    GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
Id                                                                            
1        1710             1             0         2         1             3   
2        1262             0             1  

In [5]:
X_test_full.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1461          20       RH         80.0    11622   Pave   NaN      Reg   
1462          20       RL         81.0    14267   Pave   NaN      IR1   
1463          60       RL         74.0    13830   Pave   NaN      IR1   
1464          60       RL         78.0     9978   Pave   NaN      IR1   
1465         120       RL         43.0     5005   Pave   NaN      IR1   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
Id                                                                       
1461         Lvl    AllPub    Inside       Gtl        NAmes      Feedr   
1462         Lvl    AllPub    Corner       Gtl        NAmes       Norm   
1463         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1464         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1465         HLS    AllPub    Inside       Gtl      StoneBr       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
Id                                                                         
1461       Norm     1Fam     1Story            5            6       1961   
1462       Norm     1Fam     1Story            6            6       1958   
1463       Norm     1Fam     2Story            5            5       1997   
1464       Norm     1Fam     2Story            6            6       1998   
1465       Norm   TwnhsE     1Story            8            5       1992   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                         
1461          1961     Gable  CompShg     VinylSd     VinylSd        NaN   
1462          1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace   
1463          1998     Gable  CompShg     VinylSd     VinylSd        NaN   
1464          1998     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1465          1992     Gable  CompShg     HdBoard     HdBoard        NaN   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
Id                                                                  
1461         0.0        TA        TA     CBlock       TA       TA   
1462       108.0        TA        TA     CBlock       TA       TA   
1463         0.0        TA        TA      PConc       Gd       TA   
1464        20.0        TA        TA      PConc       TA       TA   
1465         0.0        Gd        TA      PConc       Gd       TA   

     BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
Id                                                                    
1461           No          Rec       468.0          LwQ       144.0   
1462           No          ALQ       923.0          Unf         0.0   
1463           No          GLQ       791.0          Unf         0.0   
1464           No          GLQ       602.0          Unf         0.0   
1465           No          ALQ       263.0          Unf         0.0   

      BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
Id                                                                     
1461      270.0        882.0    GasA        TA          Y      SBrkr   
1462      406.0       1329.0    GasA        TA          Y      SBrkr   
1463      137.0        928.0    GasA        Gd          Y      SBrkr   
1464      324.0        926.0    GasA        Ex          Y      SBrkr   
1465     1017.0       1280.0    GasA        Ex          Y      SBrkr   

      1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
Id                                                                              
1461       896         0             0        896           0.0           0.0   
1462      1329         0             0       1329           0.0           0.0   
1463       928       701             0       1629           0.0           0.0   
1464       926   

In [6]:
# check shapes for reference
print(X_full.shape)
print(X_test_full.shape)

(1460, 80)
(1459, 79)


In [7]:
# let's split the training data into our data and target
X_full.dropna(subset='SalePrice', axis=0, inplace=True)
y = np.log(X_full['SalePrice']) # log for evaluation
X_full.drop('SalePrice', axis=1, inplace=True)

In [8]:
# we could use train_test_split, but i am going to use CV

In [9]:
# let's split up our categorical and numerical columns

ordinal_qual_cols = ['ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'HeatingQC', 'KitchenQual', 'FireplaceQu', 'GarageQual', 'GarageCond', 'PoolQC']

ordinal_cols = ordinal_qual_cols

onehot_cols = [col for col in X_full.columns if 
                   (X_full[col].dtype == 'object') and 
                   (col not in ordinal_cols)]

categorical_cols = ordinal_cols + onehot_cols

numerical_cols = [col for col in X_full.columns if 
                  (X_full[col].dtype in ['int64','float64'])]

print(categorical_cols)
print(numerical_cols)

cols = categorical_cols + numerical_cols

X_train = X_full[cols].copy()
X_test = X_test_full[cols].copy()

['ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'HeatingQC', 'KitchenQual', 'FireplaceQu', 'GarageQual', 'GarageCond', 'PoolQC', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'Foundation', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'CentralAir', 'Electrical', 'Functional', 'GarageType', 'GarageFinish', 'PavedDrive', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']
['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'Enclos

In [10]:
# let's figure out how we want to process our data
X_train.head()

ExterQual ExterCond BsmtQual BsmtCond HeatingQC KitchenQual FireplaceQu  \
Id                                                                           
1         Gd        TA       Gd       TA        Ex          Gd         NaN   
2         TA        TA       Gd       TA        Ex          TA          TA   
3         Gd        TA       Gd       TA        Ex          Gd          TA   
4         TA        TA       TA       Gd        Gd          Gd          Gd   
5         Gd        TA       Gd       TA        Ex          Gd          TA   

   GarageQual GarageCond PoolQC MSZoning Street Alley LotShape LandContour  \
Id                                                                           
1          TA         TA    NaN       RL   Pave   NaN      Reg         Lvl   
2          TA         TA    NaN       RL   Pave   NaN      Reg         Lvl   
3          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   
4          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   
5          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
Id                                                                             
1     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
2     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
3     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
4     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
5     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

   HouseStyle RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                    
1      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   
2      1Story     Gable  CompShg     MetalSd     MetalSd        NaN   
3      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4      2Story     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
5      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   Foundation BsmtExposure BsmtFinType1 BsmtFinType2 Heating CentralAir  \
Id                                                                        
1       PConc           No          GLQ          Unf    GasA          Y   
2      CBlock           Gd          ALQ          Unf    GasA          Y   
3       PConc           Mn          GLQ          Unf    GasA          Y   
4      BrkTil           No          ALQ          Unf    GasA          Y   
5       PConc           Av          GLQ          Unf    GasA          Y   

   Electrical Functional GarageType GarageFinish PavedDrive Fence MiscFeature  \
Id                                                                              
1       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
2       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
3       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
4       SBrkr        Typ     Detchd          Unf          Y   NaN         NaN   
5       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   

   SaleType SaleCondition  MSSubClass  LotFrontage  LotArea  OverallQual  \
Id                                                                         
1        WD        Normal          60         65.0     8450            7   
2        WD        Normal          20         80.0     9600            6   
3        WD        Normal          60         68.0    11250            7   
4        WD       Abnorml          70         60.0     9550            7   
5        WD        Normal          60         84.0    14260            8   

    OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  \
Id                                                                             
1             5       2003          2003       196.0         70

In [11]:
# we can likely let one hot encoding handle all of these nulls
# X_train[categorical_cols].isna().sum()

In [12]:
# LotFrontage, GarageYrBlt, and MasVnrArea only ones with na. lets look further in
# X_train[numerical_cols].isna().sum()

In [13]:
# X_train[X_train['LotFrontage'].isna()].head()
# X_train[X_train.LotFrontage == 0]
# it seems like any NA values in lotfrontage is if the house doesn't have any
# we can set this column's na values to 0 most likely

In [14]:
# X_train[X_train['GarageYrBlt'].isna()].head()
# it seems like any na values in this column are when there is no garage
# X_train.head()
# the garage dates correlate pretty well with when the house was built
# i think dropping the garage year built and instead adding a hasGarage will be best

In [15]:
# X_train[X_train['MasVnrArea'].isna()].head()
# it seems like an NaN value means there is not one, so we can input 0 for the missing values here

In [16]:
# no cardinality is too high, so no need to drop any categorical
# [X_train[col].nunique() for col in X_train[categorical_cols].columns]

In [17]:
# let's look into the testing data too to make sure these transformations are ok
# the one hot seems like it will work here, but a has pool feature should be added
# X_test[categorical_cols].isna().sum()

In [18]:
# it appears that all of these NA values would also likely be best replaced with a 0
# X_test[numerical_cols].isna().sum()

In [19]:
# the cardinalities are roughly the same, so one hot will still be the move
# [X_test[col].nunique() for col in X_test[categorical_cols].columns]

In [20]:
# now that i know what the data is like, i will start feature engineering

for df in [X_train, X_test]:
    df['TotalSF'] = (df['TotalBsmtSF'] + df['1stFlrSF'] + df['2ndFlrSF'])
    df['TotalBath'] = (df['BsmtHalfBath'] * .5 + df['BsmtFullBath'] + df['FullBath'] + df['HalfBath'] * .5)
    df['SellAge'] = (df['YrSold'] - df['YearBuilt'])
    df['OverallRating'] = (df['OverallQual'] + df['OverallCond'])

if 'TotalSF' not in numerical_cols:
    numerical_cols.extend(['TotalSF',
                           'TotalBath',
                           'SellAge',
                           'OverallRating',
                          ])

if 'GarageYrBlt' in numerical_cols:
    numerical_cols.remove('GarageYrBlt')
    numerical_cols.remove('BsmtFullBath')
    numerical_cols.remove('BsmtHalfBath')
    numerical_cols.remove('FullBath')
    numerical_cols.remove('HalfBath')

categorical_cols = ordinal_cols + onehot_cols
cols = categorical_cols + numerical_cols
X_train = X_train[cols].copy()
X_test = X_test[cols].copy()
X_train.head()

ExterQual ExterCond BsmtQual BsmtCond HeatingQC KitchenQual FireplaceQu  \
Id                                                                           
1         Gd        TA       Gd       TA        Ex          Gd         NaN   
2         TA        TA       Gd       TA        Ex          TA          TA   
3         Gd        TA       Gd       TA        Ex          Gd          TA   
4         TA        TA       TA       Gd        Gd          Gd          Gd   
5         Gd        TA       Gd       TA        Ex          Gd          TA   

   GarageQual GarageCond PoolQC MSZoning Street Alley LotShape LandContour  \
Id                                                                           
1          TA         TA    NaN       RL   Pave   NaN      Reg         Lvl   
2          TA         TA    NaN       RL   Pave   NaN      Reg         Lvl   
3          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   
4          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   
5          TA         TA    NaN       RL   Pave   NaN      IR1         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
Id                                                                             
1     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
2     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
3     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
4     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
5     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

   HouseStyle RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                    
1      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   
2      1Story     Gable  CompShg     MetalSd     MetalSd        NaN   
3      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4      2Story     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
5      2Story     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   Foundation BsmtExposure BsmtFinType1 BsmtFinType2 Heating CentralAir  \
Id                                                                        
1       PConc           No          GLQ          Unf    GasA          Y   
2      CBlock           Gd          ALQ          Unf    GasA          Y   
3       PConc           Mn          GLQ          Unf    GasA          Y   
4      BrkTil           No          ALQ          Unf    GasA          Y   
5       PConc           Av          GLQ          Unf    GasA          Y   

   Electrical Functional GarageType GarageFinish PavedDrive Fence MiscFeature  \
Id                                                                              
1       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
2       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
3       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   
4       SBrkr        Typ     Detchd          Unf          Y   NaN         NaN   
5       SBrkr        Typ     Attchd          RFn          Y   NaN         NaN   

   SaleType SaleCondition  MSSubClass  LotFrontage  LotArea  OverallQual  \
Id                                                                         
1        WD        Normal          60         65.0     8450            7   
2        WD        Normal          20         80.0     9600            6   
3        WD        Normal          60         68.0    11250            7   
4        WD       Abnorml          70         60.0     9550            7   
5        WD        Normal          60         84.0    14260            8   

    OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  \
Id                                                                             
1             5       2003          2003       196.0         70

In [21]:
# i can begin creating the pipeline now

quality_order = ['NA', 'Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_qual_transformer = Pipeline(steps=[
    ('ordqual', OrdinalEncoder(categories = [quality_order]*len(ordinal_qual_cols), handle_unknown='use_encoded_value', unknown_value=-1))
])

onehot_transformer = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

numerical_transformer = Pipeline(steps=[
    ('impute', SimpleImputer(strategy='constant', fill_value=0))
])

preprocessor = ColumnTransformer(transformers=[
    ('ordq', ordinal_qual_transformer, ordinal_qual_cols),
    ('oh', onehot_transformer, onehot_cols),
    ('num', numerical_transformer, numerical_cols)
])

In [22]:
# it is time to create the model in the pipeline

pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', XGBRegressor(#n_estimators=450, 
                           #max_depth=2, 
                           #learning_rate=.2, 
                           #colsample_bytree=.6,
                           #alpha=.1,
                           random_state=1))
])

In [23]:
# now i will use grid search to find the best parameters for the model

param_grid={
    'model__n_estimators':[621],
    'model__max_depth':[2],
    'model__learning_rate':[.125],
    'model__colsample_bytree':[.525],
    'model__alpha':[.125],
    'model__lambda':[1.1],
}

#n_estimators=621, max_depth=2, learning_rate=.125, colsample_bytree=.525, alpha=.125, lambda=1.1

grid_search = GridSearchCV(
    estimator=pipeline,
    param_grid=param_grid,
    cv=5,
    scoring='neg_root_mean_squared_error',
    n_jobs=-1,
    verbose=2
)

grid_search.fit(X_train, y)
best_pipeline = grid_search.best_estimator_
grid_search.best_params_

Fitting 5 folds for each of 1 candidates, totalling 5 fits


{'model__alpha': 0.125,
 'model__colsample_bytree': 0.525,
 'model__lambda': 1.1,
 'model__learning_rate': 0.125,
 'model__max_depth': 2,
 'model__n_estimators': 621}

In [24]:
# use cv to see how strong pipeline is

cv_scores = -1 * cross_val_score(
    best_pipeline,
    X_train,
    y,
    cv=5,
    scoring='neg_root_mean_squared_error'
)

print("RMSE per fold:", cv_scores)
print("Average RMSE:", cv_scores.mean())
# 

RMSE per fold: [0.10685894 0.12853169 0.12441899 0.10125366 0.1210169 ]
Average RMSE: 0.11641603498024163


Best submitted - 0.11641603498024163: bathrooms out of df, params n_estimators=621, max_depth=2, learning_rate=.125, colsample_bytree=.525, alpha=.125, lambda=1.1

In [25]:
# using permutation importance to see what the most/least important features are

result = permutation_importance(best_pipeline,
                                X_train,
                                y,
                                n_repeats=5,
                                random_state=1,
                                scoring='neg_root_mean_squared_error'
                               )

importance_df = pd.DataFrame({
    'feature': X_train.columns,
    'importance_mean': result.importances_mean,
    'importance_std': result.importances_std
}).sort_values(by='importance_mean', ascending=False)

print(importance_df.head(10))
print(importance_df.tail(10))

          feature  importance_mean  importance_std
74        TotalSF         0.123276        0.002408
46    OverallQual         0.049689        0.001834
77  OverallRating         0.020228        0.000678
75      TotalBath         0.018524        0.000632
45        LotArea         0.014437        0.001312
18   Neighborhood         0.013011        0.000611
10       MSZoning         0.012117        0.000649
58      GrLivArea         0.010638        0.000384
48      YearBuilt         0.009546        0.000631
49   YearRemodAdd         0.008315        0.000351
        feature  importance_mean  importance_std
23    RoofStyle         0.000065        0.000025
17    LandSlope         0.000052        0.000016
0     ExterQual         0.000026        0.000015
20   Condition2         0.000018        0.000018
21     BldgType         0.000000        0.000000
15    Utilities         0.000000        0.000000
71      MiscVal         0.000000        0.000000
11       Street         0.000000        0.00000

In [26]:
# once satisfied with the MAE, we can predict the testing data and submit
best_pipeline.fit(X_train, y)

predictions = np.exp(best_pipeline.predict(X_test)) # re-exponent from beginning log

submission = pd.DataFrame({
    'Id':X_test.index,
    'SalePrice':predictions
})
submission.to_csv('submission.csv', index=False)
pd.read_csv('submission.csv').head()

Id  SalePrice
0  1461  118045.05
1  1462  160407.95
2  1463  183783.97
3  1464  195798.28
4  1465  185839.72